### PR TITLE
MDEV-39219: buffer overflow in PROXY protocol v1 header parsing

### PR DESCRIPTION
**Pull request by uwezkhan**
This fixes a buffer overflow issue in the PROXY protocol v1 header parsing.

The loop that reads the header could fill the entire buffer and then add a null terminator past the boundary, which leads to an off-by-one overflow.

I updated the loop condition to make sure there is always space left for the null terminator.

Also added a small check to handle cases where the connection closes early, so we don’t process incomplete data.

This change does not affect normal behavior, it just makes the parsing safer.

### CODE REVIEW

File: sql/proxy_protocol.cc [hunk 1]
```diff
@@ -195,10 +195,10 @@ int parse_proxy_protocol_header(NET *net, proxy_peer_info *peer_info)
   if (have_v1_header)
   {
     /* Read until end of header (newline character)*/
-    while(pos < sizeof(hdr))
+    while(pos < sizeof(hdr) - 1)
     {
       long len= (long)vio_read(vio, hdr + pos, 1);
-      if (len < 0)
+      if (len <= 0)
```
Comment (gkodinov) [inline sql/proxy_protocol.cc:201]: this is wrong: you can get a 0 if there's no bytes to read when you call recv(). But there may be more bytes to read later. It needs to keep trying.

File: sql/proxy_protocol.cc [hunk 2]
```diff
@@ -195,10 +195,10 @@ int parse_proxy_protocol_header(NET *net, proxy_peer_info *peer_info)
   if (have_v1_header)
   {
     /* Read until end of header (newline character)*/
-    while(pos < sizeof(hdr))
+    while(pos < sizeof(hdr) - 1)
```
Comment (gkodinov) [inline sql/proxy_protocol.cc:198]: This is wrong: it MUST be able to parse the whole header. I'd increase the buffer's length and have a termination zero as an extra.

**Review by gkodinov — CHANGES_REQUESTED** [review]
Thank you for your contribution! This is a preliminary review.

First of all: sign the CLA please: either pick BSD or MariaCLA.

Secondly: Please add a commit message to your commit that's compliant with MariaDB coding standards.

Thirdly: This is a bug. Find the lowest *SUPPORTED* version this is in and target that instead of main. I'd say 10.11.

And, last but not least: please add a test! There are regression tests for the proxy protocol. Extend these.

File: sql/proxy_protocol.cc (hunk 1 above)
Comment (vaintroub) [reply sql/proxy_protocol.cc:201]: I think it was correct @gkodinov . 0 is returned from recv if and only if peer closes socket normally (FIN, not RESET), at least when reading more than 0 bytes.

Comment (gkodinov) [reply sql/proxy_protocol.cc:201]: https://man7.org/linux/man-pages/man2/recv.2.html says:

>  The value 0 may also be returned if the requested number of bytes
       to receive from a stream socket was 0.

Comment (vaintroub) [reply sql/proxy_protocol.cc:201]: yes, this does not contradict to what  I said previously . We are not reading 0 bytes. We're reading 1 byte.

Comment (gkodinov) [reply sql/proxy_protocol.cc:201]: It doesn't contradict what I said either I believe: it's wrong to treat having read 0 bytes as an error condition. If the connection is still open more might come on the next read.

Comment (vaintroub) [reply sql/proxy_protocol.cc:201]: Nope, this is wrong. 0 is an error condition. It means, other side did close(), closesocket(), or shutdown(SHUT_WR). Nothing more will come. Unless on some weird reason you using recv() with 3rd parameter 0. But nobody does it here, and I do not see any reason to ever call recv() with 3rd parameter 0, on Linux or elsewhere. If recv() trying to read 1 byte : if recv() is non-blocking, and there is nothing to read, it returns -1 with (WSA)EWOULDBLOCK. if recv() is blocking, and there is nothing to read it will block until it can read 1 byte, or until error, and return code 0 is this specific error (peer closed/shut down connection orderly), and -1 is for other errors (ECONNRESET, ETIMEDOUT).

Comment (uwezkhan) [reply sql/proxy_protocol.cc:201]: Thanks for the discussion, this is helpful.

From what I understand, since we’re reading 1 byte in blocking mode, a return value of 0 would indicate that the peer has closed the connection and no more data will arrive. In that case, continuing to read would likely result in a loop without progress.

To keep things safe and avoid partial parsing or infinite loops, I’m thinking of treating len <= 0 as a termination condition here.

That said, I want to make sure I’m aligned with the expected behavior in this part of the codebase. If there’s a preferred pattern for handling this in MariaDB networking code, I can follow that.

For the buffer issue, I’ll update the fix to increase the buffer size by one byte and keep full header parsing intact with safe null termination.

Let me know if that approach sounds good and I’ll update the patch accordingly.

Comment (vaintroub) [reply sql/proxy_protocol.cc:201]: I think this is pretty good, 0 is in this case definitely the "close" condition. If you continue to try to read past that, I do not know for sure,  I *think* at some point it could return -1 with ECONNRESET, or EPIPE or something like that.  

You can give it a try. The tests for proxy protocol are in tests/mysql_client_test.c  (look for `test_proxy` in the file), you can add yours, with header that would cause buffer overflow.

A test with an incomplete header and shutdown(SHUT_WR) (half-closing connection, which I'd choose as a testable version for the "close" scenario) is trickier, as you can't use client API for that. I.e you'd need just a TCP socket that connects to the server, sends incomplete proxy header, half-closes connection with shutdown(SHUT_WR), reads server "welcome" packet, reads and checks for expected server error packet, then closesocket().

**Comment by uwezkhan** [comment]
Thanks for the review and guidance!

I’ve updated the patch to:
- use correct read semantics (pos += len)
- ensure safe buffer handling with full header support
- handle connection close safely during read

I’ve also added and registered a regression test for maximum header length in mysql_client_test.c.

Please let me know if anything else needs adjustment.

**Comment by vaintroub** [comment]
There is a (arguably incorrect) error message from GCC
/home/buildbot/tests/mysql_client_test.c:20764:3: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
20764 |   const char *prefix = "PROXY TCP4 1.2.3.4 5.6.7.8 1234 5678";

eventhough we compile with C99, it still throughs this "ISO C90" warning, which turns to error in Linux debug compilation. Anyway, can you move that variable declaration at the start of the block, to fix the GCC.

**Comment by uwezkhan** [comment]
Thanks for pointing that out.

I’ve moved the variable declarations to the beginning of the function to comply with the C90-style requirement and avoid the GCC warning.

I’ve pushed the updated patch.

**Review by gkodinov — CHANGES_REQUESTED** [review]
Thanks for working on this. I'll leave the finer points to the final reviewer.

Please:
* squash your commit into a single one
* rebase on 10.11
* Add a commit message to your commit that's compliant with the mariadb coding standards.

**Comment by gkodinov** [comment]
There was no reply to my preliminary review from couple of weeks ago. Moving this to "draft" state. Please move back to "open" when/if you intend to keep working on it.

**Comment by uwezkhan** [comment]
I recreated this PR on top of 10.11 with the commits squashed into a single commit and a MariaDB-compliant commit message.

Updated PR: https://github.com/MariaDB/server/pull/4998

Closing this one in favor of the updated version.
